govcd/system.go

```diff
@@ -43,6 +59,213 @@ func CreateOrg(vcdClient *VCDClient, name string, fullName string, description s
+
+	if egwConfiguration.Name != egwName {
+		return Task{}, fmt.Errorf("name mismatch: '%s' used as parameter but '%s' in the configuration structure", egwName, egwConfiguration.Name)
+	}
+	adminOrg, err := GetAdminOrgByName(vcdClient, orgName)
+	if err != nil {
+		return Task{}, err
+	}
+	vdc, err := adminOrg.GetVdcByName(vdcName)
+	if err != nil {
+		return Task{}, err
+	}
+
+	egwCreateHREF := vcdClient.Client.VCDHREF
+
+	ID, err := getBareEntityUuid(vdc.Vdc.ID)
+	if err != nil {
+		return Task{}, fmt.Errorf("error retrieving ID from Vdc %s: %s", vdcName, err)
+	}
+	if ID == "" {
+		return Task{}, fmt.Errorf("error retrieving ID from Vdc %s - empty ID returned", vdcName)
+	}
+	egwCreateHREF.Path += fmt.Sprintf("/admin/vdc/%s/edgeGateways", ID)
+
+	// The first task is the creation task. It is quick, and does only create the vCD entity,
+	// but not yet deploy the underlying VM
+	creationTask, err := vcdClient.Client.ExecuteTaskRequest(egwCreateHREF.String(), http.MethodPost,
+		"application/vnd.vmware.admin.edgeGateway+xml", "error instantiating a new Edge Gateway: %s", egwConfiguration)
+
+	if err != nil {
+		return Task{}, err
+	}
+
+	err = creationTask.WaitTaskCompletion()
+
+	if err != nil {
+		return Task{}, err
+	}
+
+	// After creation, there is a build task that supervises the gateway deployment
+	for _, innerTask := range creationTask.Task.Tasks.Task {
+		if innerTask.OperationName == "networkEdgeGatewayCreate" {
+			deployTask := Task{
+				Task:   innerTask,
+				client: &vcdClient.Client,
+			}
+			return deployTask, nil
+		}
+	}
+	return Task{}, fmt.Errorf("no deployment task found for edge gateway %s - The edge gateway might have been created, but not deployed properly", egwName)
+}
+
+func createEdgeGateway(vcdClient *VCDClient, egwc EdgeGatewayCreation, egwConfiguration *types.EdgeGateway) (EdgeGateway, error) {
+	var task Task
+	var err error
+	if egwConfiguration != nil {
+		task, err = CreateAndConfigureEdgeGatewayAsync(vcdClient, egwc.OrgName, egwc.VdcName, egwc.Name, egwConfiguration)
+	} else {
+		task, err = CreateEdgeGatewayAsync(vcdClient, egwc)
+	}
+
+	if err != nil {
+		return EdgeGateway{}, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return EdgeGateway{}, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	// The edge gateway is created. Now we retrieve it from the server
+	org, err := GetAdminOrgByName(vcdClient, egwc.OrgName)
+	if err != nil {
+		return EdgeGateway{}, err
+	}
+	vdc, err := org.GetVdcByName(egwc.VdcName)
+	if err != nil {
+		return EdgeGateway{}, err
+	}
+	egw, err := vdc.FindEdgeGateway(egwc.Name)
+	if err != nil {
+		return EdgeGateway{}, err
+	}
+	return egw, nil
+}
+
+// CreateAndConfigureEdgeGateway creates an edge gateway using a full configuration structure
+func CreateAndConfigureEdgeGateway(vcdClient *VCDClient, orgName, vdcName, egwName string, egwConfiguration *types.EdgeGateway) (EdgeGateway, error) {
```

Comment:
This would create a mismatch where we need to use the result of this function in combination with other functions that return a structure containing pointers (the default in all govcd). Either we change all of them (meaning we change the type) or we leave them as they are.
In my case, I would need to deal with structure and pointer within adjacent statements, and I'd like to avoid this situation.

Reply:
Ok. This means not this time at least.